lib/stream_data.ex

```diff
@@ -496,44 +506,43 @@ defmodule StreamData do
   Shrinks towards with the smallest absolute value that still lie in `range`.
   """
   @spec integer(Range.t()) :: t(integer())
-  def integer(_lower.._upper = range) do
+  def integer(lower..upper = _range) do
     new(fn seed, _size ->
-      range
-      |> uniform_in_range(seed)
-      |> integer_lazy_tree(range)
+      {init, _next_seed} = uniform_in_range(lower, upper, seed)
+      integer_lazy_tree(init, lower, upper)
     end)
   end
 
-  defp integer_lazy_tree(int, range) do
-    lazy_tree(int, &integer_lazy_tree(int, range, _current = int, &1, &2))
+  defp integer_lazy_tree(int, lower, upper) do
+    lazy_tree(int, &integer_lazy_tree(int, lower, upper, _current = int, &1, &2))
   end
 
-  defp integer_lazy_tree(_int, _range, _current, {:halt, acc}, _fun) do
+  defp integer_lazy_tree(_int, _lower, _upper, _current, {:halt, acc}, _fun) do
     {:halted, acc}
   end
 
-  defp integer_lazy_tree(int, range, current, {:suspend, acc}, fun) do
-    {:suspended, acc, &integer_lazy_tree(int, range, current, &1, fun)}
+  defp integer_lazy_tree(int, lower, upper, current, {:suspend, acc}, fun) do
+    {:suspended, acc, &integer_lazy_tree(int, lower, upper, current, &1, fun)}
   end
 
-  defp integer_lazy_tree(int, range, current, {:cont, acc}, fun) do
-    to_emit = int - current
-
-    cond do
-      to_emit == int ->
+  defp integer_lazy_tree(int, lower, upper, current, {:cont, acc}, fun) do
+    case int - current do
+      ^int ->
         {:done, acc}
 
-      to_emit in range ->
-        lazy_tree = integer_lazy_tree(to_emit, range)
-        integer_lazy_tree(int, range, div(current, 2), fun.(lazy_tree, acc), fun)
+      to_emit when to_emit >= lower and to_emit <= upper ->
```

Comment:
So we take care of `lower > upper` in `uniform_in_range/3`, but I can't see that we do here as well? If `lower > upper` here, I don't think this will work?

Reply:
This was addressed by always making sure the range is in the right order before entering the generation loop